cpp/src/join/quadtree_point_in_polygon.cu

```diff
@@ -0,0 +1,573 @@
+  auto d_poly_offsets  = cudf::column_device_view::create(poly_offsets, stream);
+  auto d_ring_offsets  = cudf::column_device_view::create(ring_offsets, stream);
+  auto d_poly_points_x = cudf::column_device_view::create(poly_points_x, stream);
+  auto d_poly_points_y = cudf::column_device_view::create(poly_points_y, stream);
+
+  quad_pip_phase1_kernel<T><<<num_unit_pairs, threads_per_block, 0, stream>>>(d_pq_poly_idx.data(),
+                                                                              d_pq_quad_idx.data(),
+                                                                              d_quad_offset.data(),
+                                                                              d_quad_length.data(),
+                                                                              *d_quad_offsets,
+                                                                              *d_point_indices,
+                                                                              *d_point_x,
+                                                                              *d_point_y,
+                                                                              *d_poly_offsets,
+                                                                              *d_ring_offsets,
+                                                                              *d_poly_points_x,
+                                                                              *d_poly_points_y,
+                                                                              d_num_hits.data());
+
+  CUDA_TRY(cudaStreamSynchronize(stream));
+
+  // remove poly-quad pair with zero hits
+  auto valid_pq_pair_iter = make_zip_iterator(d_pq_poly_idx.begin(),
+                                              d_pq_quad_idx.begin(),
+                                              d_quad_offset.begin(),
+                                              d_quad_length.begin(),
+                                              d_num_hits.begin());
+
+  uint32_t num_valid_pair = thrust::distance(
+    valid_pq_pair_iter,
+    thrust::remove_if(rmm::exec_policy(stream)->on(stream),
+                      valid_pq_pair_iter,
+                      valid_pq_pair_iter + num_unit_pairs,
+                      valid_pq_pair_iter,
+                      [] __device__(auto const &v) { return thrust::get<4>(v) == 0; }));
+
+  d_pq_poly_idx.resize(num_valid_pair, stream);
+  d_pq_quad_idx.resize(num_valid_pair, stream);
+  d_quad_offset.resize(num_valid_pair, stream);
+  d_quad_length.resize(num_valid_pair, stream);
+  d_num_hits.resize(num_valid_pair, stream);
+
+  uint32_t total_hits =
+    thrust::reduce(rmm::exec_policy(stream)->on(stream), d_num_hits.begin(), d_num_hits.end());
+
+  // prefix sum on numbers to generate offsets
+  thrust::exclusive_scan(
+    rmm::exec_policy(stream)->on(stream), d_num_hits.begin(), d_num_hits.end(), d_num_hits.begin());
```

Comment:
Doing a reduction and a prefix sum on the same array is overkill, you can get the the scan and the sum in one go.  If you make `d_num_hits` one element larger than needed initially (and operate on n-1 elements everywhere above):

```suggestion
  // prefix sum on numbers to generate offsets
  thrust::exclusive_scan(
    rmm::exec_policy(stream)->on(stream), d_num_hits.begin(), d_num_hits.end(), d_num_hits.begin());
  uint32_t total_hits = d_num_hits.back_element(stream);
```
Note `back_element` calls `element`, which synchronizes the stream so `total_hits` is valid and ready when it returns.